File: frontend-next/lib/redact.ts
/**
 * PII redaction — TypeScript port of swarmtrace/redact.py.
 *
 * Used by the /api/ingest and /api/events edge routes to scrub PII from
 * args/output/error BEFORE the row hits Supabase. This is defense-in-depth:
 * the SDK already redacts (swarmtrace/redact.py) before sending, but any
 * client that posts directly (curl, MCP, a third-party SDK port) bypasses
 * the SDK. Redacting at the ingest boundary means PII never lands in the
 * DB regardless of which client sent it.
 *
 * Pure functions — no I/O, no global state, no exceptions. Safe to call
 * from the edge hot path.
 *
 * Categories scrubbed (identical to the Python implementation):
 *   1. Emails (RFC-ish local@domain.tld)
 *   2. API-key-shaped strings (sk-, sk-ant-, ghp_, github_pat_, xox[bpoa]-,
 *      AKIA, sk_live_/sk_test_, rk_live_, pypi-AgEI, AIza — with length gates)
 *   3. Credit card numbers (13–19 digit groups, Luhn-checked — NOT bare regex,
 *      so 16-digit trace IDs / UUID fragments pass through)
 *   4. JWTs (eyJ….….… three-segment base64url shape)
 *
 * The regexes and Luhn algorithm are line-for-line ports of redact.py.
 * The Python test suite (tests/test_redact.py, 48 tests) is the source of
 * truth for the contract; the TS test suite (scripts/test-redact.mjs)
 * mirrors the critical cases.
 */

const REDACTED = '[REDACTED]'

// ── Patterns (identical to redact.py) ────────────────────────────────────────

const EMAIL_RE = /\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b/g

const API_KEY_RE = new RegExp(
  '\\b(' +
    'sk-ant-[A-Za-z0-9_\\-]{20,}' +
    '|sk-[A-Za-z0-9_\\-]{20,}' +
    '|gh[pousr]_[A-Za-z0-9]{36,}' +
    '|github_pat_[A-Za-z0-9_]{82,}' +
    '|xox[bpoa]-[A-Za-z0-9\\-]{10,}' +
    '|AKIA[0-9A-Z]{16}' +
    '|sk_(?:live|test)_[A-Za-z0-9]{24,}' +
    '|rk_live_[A-Za-z0-9]{24,}' +
    '|pypi-AgEI[A-Za-z0-9_\\-]{20,}' +
    '|AIza[0-9A-Za-z_\\-]{35}' +
    ')',
  'g',
)

const JWT_RE = /\beyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\b/g

// Candidate credit card numbers — runs of 13–19 digits with optional
// single space or dash separators. Each candidate is Luhn-checked.
const CC_CANDIDATE_RE = /\b(?:\d[ -]?){13,19}\b/g

// ── Luhn check (port of luhn_ok) ────────────────────────────────────────────

export function luhnOk(digits: string): boolean {
  if (!digits || !/^\d+$/.test(digits)) return false
  if (digits.length < 13 || digits.length > 19) return false
  let total = 0
  const parity = digits.length % 2
  for (let i = 0; i < digits.length; i++) {
    let d = parseInt(digits[i], 10)
    if (i % 2 === parity) {
      d *= 2
      if (d > 9) d -= 9
    }
    total += d
  }
  return total % 10 === 0
}

// ── Credit card redactor (Luhn-gated) ───────────────────────────────────────

function redactCreditCards(text: string): string {
  return text.replace(CC_CANDIDATE_RE, (raw) => {
    const digits = raw.replace(/[^0-9]/g, '')
    return luhnOk(digits) ? REDACTED : raw
  })
}

// ── Public entry point ──────────────────────────────────────────────────────

export function redact(text: string | null | undefined): string | null {
  if (text === null || text === undefined) return null
  if (typeof text !== 'string') text = String(text)
  if (text === '') return text
  // Order matters slightly: emails first (so the @ doesn't become part of a
  // JWT-like sequence), then API keys, then JWTs, then credit cards.
  text = text.replace(EMAIL_RE, REDACTED)
  text = text.replace(API_KEY_RE, REDACTED)
  text = text.replace(JWT_RE, REDACTED)
  text = redactCreditCards(text)
  return text
}

/** Split snake/kebab/camel-case keys into normalized words. */
function keyWords(key: string): string[] {
  return key
    .replace(/([a-z0-9])([A-Z])/g, '$1_$2')
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter(Boolean)
}

const SENSITIVE_KEY_WORDS = new Set([
  'password', 'passwd', 'pwd', 'secret', 'token', 'authorization', 'auth',
  'cookie', 'session', 'credential', 'otp', 'totp', 'mfa', 'csrf',
  'ssn', 'cvv', 'cvc', 'pin',
])

function isSensitiveKey(key: string): boolean {
  const words = keyWords(key)
  // Length/count metadata is safe and useful even when it describes a secret.
  if (['chars', 'length', 'count', 'size'].includes(words.at(-1) ?? '')) return false
  if (words.some(word => SENSITIVE_KEY_WORDS.has(word))) return true
  const joined = words.join('_')
  return [
    'api_key', 'access_key', 'private_key', 'client_secret',
    'recovery_code', 'backup_code', 'security_answer',
  ].some(part => joined.includes(part))
}

function isUrlKey(key: string): boolean {
  const words = keyWords(key)
  const last = words.at(-1)
  return last === 'url' || last === 'uri' || last === 'href'
}

/** Remove query strings and fragments, which commonly contain credentials. */
export function redactUrl(url: string): string {
  const query = url.indexOf('?')
  const fragment = url.indexOf('#')
  const indexes = [query, fragment].filter(index => index >= 0)
  const cut = indexes.length ? Math.min(...indexes) : url.length
  return url.slice(0, cut)
}

/**
 * Recursively redact strings in an object/array structure.
 *
 * In addition to pattern matching, values under credential-shaped keys are
 * removed contextually and URL-shaped fields lose their query/fragment.
 * Cycles and excessive nesting are replaced with [REDACTED] rather than
 * overflowing the server stack.
 */
export function redactDeep<T>(value: T): T {
  const seen = new WeakSet<object>()
  const MAX_DEPTH = 64

  function walk(current: unknown, depth: number, key?: string): unknown {
    if (depth > MAX_DEPTH) return REDACTED
    if (typeof current === 'string') {
      const text = key && isUrlKey(key) ? redactUrl(current) : current
      return redact(text)
    }
    if (current === null || typeof current !== 'object') return current
    if (seen.has(current)) return REDACTED

    seen.add(current)
    let result: unknown
    if (Array.isArray(current)) {
      result = current.map(item => walk(item, depth + 1))
    } else {
      const out: Record<string, unknown> = {}
      for (const [childKey, childValue] of Object.entries(current)) {
        const cleaned = isSensitiveKey(childKey)
          ? REDACTED
          : walk(childValue, depth + 1, childKey)
        // Avoid __proto__ assignment changing the result object's prototype.
        Object.defineProperty(out, childKey, {
          value: cleaned,
          enumerable: true,
          configurable: true,
          writable: true,
        })
      }
      result = out
    }
    seen.delete(current)
    return result
  }

  return walk(value, 0) as T
}

const EVENT_VALUE_METHODS = new Set(['fill', 'type', 'press', 'select_option'])

/**
 * Apply schema-aware event redaction after generic deep redaction.
 *
 * This protects direct /api/events clients that do not use the Python SDK:
 * browser value arguments are removed, browser/HTTP URLs are stripped, and
 * streamed token content is never retained chunk-by-chunk.
 */
export function redactEventData(eventType: string, value: unknown): unknown {
  const cleaned = redactDeep(value)
  if (
    value === null || typeof value !== 'object' || Array.isArray(value) ||
    cleaned === null || typeof cleaned !== 'object' || Array.isArray(cleaned)
  ) {
    return cleaned
  }

  const raw = value as Record<string, unknown>
  const out = cleaned as Record<string, unknown>
  const method = typeof raw.method === 'string' ? raw.method.toLowerCase() : ''

  if (eventType === 'browser' && Array.isArray(raw.args) && Array.isArray(out.args)) {
    const args = [...out.args]
    if (EVENT_VALUE_METHODS.has(method) && raw.args.length > 1) {
      const rawValue = String(raw.args[1] ?? '')
      args[1] = `[REDACTED(len=${rawValue.length})]`
      if (typeof out.error === 'string' && rawValue) {
        out.error = out.error.replaceAll(rawValue, args[1] as string)
      }
    }
    if (method === 'goto' && typeof raw.args[0] === 'string') {
      args[0] = redactUrl(raw.args[0])
    }
    out.args = args
  }

  if (eventType === 'llm_token') {
    if ('token' in raw) out.token = REDACTED
    if ('accumulated' in raw) out.accumulated = REDACTED
  }

  return out
}

File: swarmtrace/redact.py
"""PII redaction — pure functions, no dependencies.

Used by :func:`swarmtrace.tracer._flush` to scrub PII from ``args``,
``output``, and ``error`` *before* they hit either the local SQLite DB
or the remote ingest endpoint.  Redacting in ONE place (the ``_flush``
call site) means local and remote never disagree about what was stored.

Pure functions
--------------
Every function here takes a string (or ``None``) and returns a string
(or ``None``).  No I/O, no global state, no exceptions, no imports beyond
the standard library.  This keeps the module unit-testable in isolation
and safe to call from the hot tracing path.

Categories scrubbed
-------------------
1. **Emails** — RFC-ish ``local@domain.tld`` shape.
2. **API-key-shaped strings** — well-known provider prefixes
   (``sk-``, ``sk-ant-``, ``ghp_``, ``xox[bpoa]-``, ``AKIA``, ``AIza``,
   ``pypi-AgEI``, ``sk_live_``, ``sk_test_``, ``rk_live_``) followed by
   a length gate so a bare ``"sk-"`` doesn't match.
3. **Credit card numbers** — digit groups (13–19 digits, separators
   allowed) that pass the **Luhn check**.  A bare regex would also catch
   16-digit trace IDs, UUID fragments, and numeric IDs; Luhn is what
   separates real card numbers from those.
4. **JWTs** — three base64url segments ``eyJ….….…`` (the ``eyJ`` prefix
   is ``{"`` base64-encoded, so every JWT header starts with it).

What is NOT scrubbed
--------------------
- Phone numbers (too locale-dependent, false-positive prone).
- SSNs (region-specific; add if/when a customer asks).
- Street addresses (no compact regex shape).
- Generic "long random strings" (no reliable signal — would either
  over-match API nonces or under-match custom tokens).
"""

from __future__ import annotations

import re
from typing import overload

__all__ = ["luhn_ok", "redact"]


# --------------------------------------------------------------------------
# Patterns
# --------------------------------------------------------------------------

# Email — local part allows dots, +, -, _, %, alnum.  Domain is alnum + dots,
# TLD ≥ 2 alpha chars.  Good enough for the realistic shapes that show up in
# LLM prompts and tool args; we're not trying to validate per-RFC-5322.
_EMAIL_RE = re.compile(
    r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b"
)

# API key prefixes — curated set of well-known provider formats.  Each
# alternative has a length gate after the prefix so a 3-char "sk-" alone
# can't match.  Alternatives, in order of specificity (longest prefix first
# so e.g. ``sk-ant-`` wins over ``sk-``):
#
#   sk-ant-[A-Za-z0-9_\-]{20,}      Anthropic
#   sk-[A-Za-z0-9_\-]{20,}          OpenAI / OpenAI-compatible
#   gh[pousr]_[A-Za-z0-9]{36,}      GitHub PAT (classic — 40 chars total)
#   github_pat_[A-Za-z0-9_]{82,}    GitHub fine-grained (93 chars total)
#   xox[bpoa]-[A-Za-z0-9\-]{10,}    Slack
#   AKIA[0-9A-Z]{16}                AWS access key id (exact 20 chars)
#   sk_(?:live|test)_[A-Za-z0-9]{24,}  Stripe secret key
#   rk_live_[A-Za-z0-9]{24,}        Stripe restricted key
#   pypi-AgEI[A-Za-z0-9_\-]{20,}    PyPI API token
#   AIza[0-9A-Za-z_\-]{35}          Google API key (exact 39 chars)
#
# Length gates are deliberately generous — false negatives (missing a real
# key) are worse than false positives (redacting a 36-char alphanumeric
# string that happens to start with ``ghp_`` but isn't actually a PAT).
_API_KEY_RE = re.compile(
    r"\b("
    r"sk-ant-[A-Za-z0-9_\-]{20,}"
    r"|sk-[A-Za-z0-9_\-]{20,}"
    r"|gh[pousr]_[A-Za-z0-9]{36,}"
    r"|github_pat_[A-Za-z0-9_]{82,}"
    r"|xox[bpoa]-[A-Za-z0-9\-]{10,}"
    r"|AKIA[0-9A-Z]{16}"
    r"|sk_(?:live|test)_[A-Za-z0-9]{24,}"
    r"|rk_live_[A-Za-z0-9]{24,}"
    r"|pypi-AgEI[A-Za-z0-9_\-]{20,}"
    r"|AIza[0-9A-Za-z_\-]{35}"
    r")"
)

# JWT — three base64url segments separated by dots.  Header always starts
# with ``eyJ`` because that's ``{"`` base64-encoded.  Standard JWT shape;
# we don't try to decode/verify, just match the shape and redact.
_JWT_RE = re.compile(
    r"\beyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\b"
)

# Candidate credit card numbers — a solid run of 13-19 digits, OR digits
# grouped with a single *uniform* separator (space or dash) in one of the
# standard PAN groupings: 4-4-4-{1..7} (Visa/Mastercard/Discover/etc.,
# covers 13-19 total digits) or 4-6-5 (Amex, 15 digits). Every alternative
# starts and ends on a digit — never on a separator — so a match can never
# eat a trailing space/dash or bleed into an adjacent unrelated digit
# (previously `(?:\d[ -]?){13,19}` could consume one extra char past the
# number, silently deleting data, and could fuse two space-separated cards
# into a single match).
_CC_CANDIDATE_RE = re.compile(
    r"\b\d{13,19}\b"
    r"|\b\d{4}([ -])\d{6}\1\d{5}\b"
    r"|\b\d{4}([ -])\d{4}\2\d{4}\2\d{1,7}\b"
)


_REDACTED = "[REDACTED]"


# --------------------------------------------------------------------------
# Luhn check (public — exported for tests + reuse)
# --------------------------------------------------------------------------

def luhn_ok(digits: str) -> bool:
    """Return ``True`` iff *digits* passes the Luhn checksum.

    *digits* must contain only digit characters (callers typically strip
    spaces/dashes first).  Length must be 13–19 — the ISO/IEC 7812 range
    for payment card primary account numbers.  Outside that range, Luhn
    is meaningless, so we return ``False`` rather than running the math.

    The algorithm: starting from the leftmost digit, double every
    second digit (if the result is > 9, subtract 9), sum all digits,
    and check the total is divisible by 10.  This is the standard
    Luhn formula used by every major card network (Visa, Mastercard,
    Amex, Discover, JCB, UnionPay).
    """
    if not digits or not digits.isdigit():
        return False
    if len(digits) < 13 or len(digits) > 19:
        return False
    total = 0
    parity = len(digits) % 2
    for i, ch in enumerate(digits):
        d = int(ch)
        if i % 2 == parity:
            d *= 2
            if d > 9:
                d -= 9
        total += d
    return total % 10 == 0


# --------------------------------------------------------------------------
# Per-category redactors
# --------------------------------------------------------------------------

def _redact_credit_cards(text: str) -> str:
    """Replace Luhn-valid digit groups with ``[REDACTED]``.

    Non-Luhn candidates (e.g. a 16-digit trace ID, a 13-digit order
    number) pass through unchanged.  This is the deliberate distinction
    from a bare-regex approach — the user explicitly asked for
    Luhn-checked, not bare regex, so non-card numbers don't get
    accidentally scrubbed.
    """
    def replace(m: re.Match[str]) -> str:
        raw = m.group(0)
        digits = re.sub(r"[^0-9]", "", raw)
        return _REDACTED if luhn_ok(digits) else raw
    return _CC_CANDIDATE_RE.sub(replace, text)


# --------------------------------------------------------------------------
# Public entry point
# --------------------------------------------------------------------------

@overload
def redact(text: str) -> str: ...


@overload
def redact(text: None) -> None: ...


def redact(text: str | None) -> str | None:
    """Scrub emails, API keys, credit card numbers, and JWTs from *text*.

    Overloaded so ``None`` in means ``None`` out and ``str`` in means ``str``
    out. Without the overloads the declared return is ``str | None`` for every
    caller, and the two call sites that slice the result directly —
    ``redact(str(x))[:32000]`` in ``otlp_mapping`` and the ``args``/``output``/
    ``error`` reassignments in ``tracer._flush`` — are unprovable to a type
    checker even though they are safe at runtime. The overloads make that
    safety checkable rather than assumed.

    - ``None`` passes through unchanged (so ``tracer._flush`` can call
      ``redact(output)`` even when output is ``None``).
    - Non-string inputs are coerced via ``str()`` first.
    - Empty strings return as empty strings.
    - Idempotent: redacting an already-redacted string is a no-op
      (``[REDACTED]`` doesn't match any of the four patterns).

    The order of operations matters slightly: emails first (so the
    ``@`` doesn't accidentally become part of a JWT-like sequence),
    then API keys (longest prefix first within the alternation), then
    JWTs, then credit cards (Luhn-checked).  In practice the four
    categories don't overlap, but the order is defensive.
    """
    if text is None:
        return None
    if not isinstance(text, str):
        text = str(text)
    if not text:
        return text
    text = _EMAIL_RE.sub(_REDACTED, text)
    text = _API_KEY_RE.sub(_REDACTED, text)
    text = _JWT_RE.sub(_REDACTED, text)
    text = _redact_credit_cards(text)
    return text
